predicators/nsrt_learning/strips_learning/effect_search_learner.py

```diff
@@ -21,160 +21,121 @@
-        for (spec, add_atoms, keep_atoms) in self:
-            opt = spec[0].name
-            opt_args = ", ".join([str(a) for a in spec[1]])
-            s += f"\n{opt}({opt_args}): {add_atoms}, {keep_atoms}"
-        s += "\n"
-        return s
-
-    def add(self, option_spec: OptionSpec, add_effects: Set[LiftedAtom],
-            keep_effects: Set[LiftedAtom]) -> _EffectSets:
-        """Create a new _EffectSets with this new entry added to existing."""
-        param_option = option_spec[0]
-        new_param_option_to_groups = {
-            p: [(s, set(aa), set(ka)) for s, aa, ka in group]
-            for p, group in self._param_option_to_groups.items()
-        }
-        if param_option not in new_param_option_to_groups:
-            new_param_option_to_groups[param_option] = []
-        new_param_option_to_groups[param_option].append(
-            (option_spec, add_effects, keep_effects))
-        return _EffectSets(new_param_option_to_groups)
-
-    def remove(self, option_spec: OptionSpec, add_effects: Set[LiftedAtom],
-               keep_effects: Set[LiftedAtom]) -> _EffectSets:
-        """Create a new _EffectSets with this entry removed from existing."""
-        param_option = option_spec[0]
-        assert param_option in self._param_option_to_groups
-        new_param_option_to_groups = {
-            p: [(s, set(aa), set(ka)) for s, aa, ka in group]
-            for p, group in self._param_option_to_groups.items()
-        }
-        new_param_option_to_groups[param_option].remove(
-            (option_spec, add_effects, keep_effects))
-        return _EffectSets(new_param_option_to_groups)
+def _update_pnad_seg_to_keep_effs(pnad: PartialNSRTAndDatastore,
+                                  necessary_image: Set[GroundAtom],
+                                  ground_op: _GroundSTRIPSOperator,
+                                  obj_to_var: Dict[Object, Variable],
+                                  segment: Segment) -> None:
+    """Updates the pnad's seg_to_keep_effs_sub dictionary, which is necesssary
+    for correctly grounding keep effects to data."""
+    # Every atom in the necessary_image that wasn't in the
+    # ground_op's add effects is a possible keep effect. This
+    # may add new variables, whose mappings for this segment
+    # we keep track of in the seg_to_keep_effects_sub dict.
+    for atom in necessary_image - ground_op.add_effects:
+        keep_eff_sub = {}
+        for obj in atom.objects:
+            if obj in obj_to_var:
+                continue
+            new_var = utils.create_new_variables([obj.type],
+                                                 obj_to_var.values())[0]
+            obj_to_var[obj] = new_var
+            keep_eff_sub[new_var] = obj
+        pnad.poss_keep_effects.add(atom.lift(obj_to_var))
+        if segment not in pnad.seg_to_keep_effects_sub:
+            pnad.seg_to_keep_effects_sub[segment] = {}
+        pnad.seg_to_keep_effects_sub[segment].update(keep_eff_sub)
```

Comment:
maybe we can make this function live somewhere else instead of just here at the top of the file? I think the `BackchainingSTRIPSLearner` also basically does this, so maybe we can make it common by putting it in the `GeneralToSpecificSTRIPSLearner` class? TBD

Reply:
sounds like a good idea, and worth doing in this PR